[3] Add ReductionOverNeighbor - Seperate

lukasm91 opened the pull request:
See https://github.com/MeteoSwiss-APN/dawn/pull/269

Commit:
include prettydump

Commit:
Merge branch 'mstr' into prettydump

Commit:
in memory creation

Commit:
Merge remote-tracking branch 'mch/master' into prettydump

Commit:
Merge branch 'prettydump' into mstr

Commit:
fixes [WIP]

Commit:
done

Commit:
prototying first staement

Commit:
make it work

Commit:
add missing diffusion statements

Commit:
Fixes

Commit:
fixes for diffusion

Commit:
add += in the end

Commit:
Merge remote-tracking branch 'twicki/prototyping' into prototyping

Commit:
Fix new statement

Commit:
reduction is expr

Commit:
init can also be an expr

Commit:
add prototyping library

Commit:
Merge remote-tracking branch 'upstream/master' into prototyping

Commit:
fix

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
add iir builder

Commit:
add a seperate test and revert OptimizerContext

Commit:
Merge remote-tracking branch 'upstream/master' into prototyping

Commit:
fix merge

Commit:
fix storage types

Commit:
add unary operator

Commit:
reduce some code and add more binary/unary ops

Commit:
add make_localvar and fix stage extents

Commit:
fix warning / add todo

Commit:
local var access works

Commit:
cleanup

Commit:
abbreviate function names

Commit:
add ifstmt and conditional

Commit:
remove redundant StatementAccessesPiar

Commit:
add block stmt and some fixes

Commit:
clean up 1

Commit:
cleanup 2

Commit:
Cleanup 3

Commit:
Merge remote-tracking branch 'upstream/master' into prototyping

Commit:
Merge branch 'in-memory-iir' into prototyping

Commit:
Merge branch 'in-memory-iir' into prototyping

Commit:
fix

Commit:
revert

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
Merge branch 'prototype-compare2' into prototyping

Commit:
Merge branch 'prototype-compare2' into prototyping

Commit:
Merge branch 'prototyping-compare' into prototyping

havogt commented on src/dawn/AST/ASTExpr.cpp:
```diff
@@ -430,5 +431,33 @@ bool LiteralAccessExpr::equals(const Expr* other) const {
+ReductionOverNeighborExpr::ReductionOverNeighborExpr(std::string const& op,
+                                                     std::shared_ptr<Expr> rhs,
+                                                     std::shared_ptr<Expr> init, SourceLocation loc)
+    : Expr(EK_ReductionOverNeighborExpr, loc), op_(op), rhs_(std::move(rhs)),
```
I think the moves disable copy-elision.

lukasm91 replied:
I am not sure (I know it's a problem in the return, bu t not sure about constructores). As this was copied from another place, I replaced it with `const&` now

havogt replied:
I verified in a small snippet that copy elision is indeed disabled in this case. So the move should be avoided.

havogt replied:
In a different branch? Or didn't push?

havogt replied:
Ah clicked on view changes and was sent to the old state.

Commit:
Merge branch 'prototyping-compare' into prototyping

havogt commented on src/dawn/AST/ASTStringifier.cpp:
```diff
@@ -143,6 +143,11 @@ class StringVisitor : public ASTVisitor {
   }
+  void visit(const std::shared_ptr<ReductionOverNeighborExpr>& expr) override {
+    if(scopeDepth_ == 0)
+      ss_ << std::string(curIndent_, ' ');
```
How does it related to scopeDepth==0?

lukasm91 replied:
I have no idea, what and how this is relevant for Stmts, but obviously it is not for Expr. THese lines were outdated anyway. Fixed.

lukasm91 committed:
Unstructured backend (#285)

Adds an unstructured backend  `cxxnaiveico` as a copy of the naive backend `cxxnaive`

Commit:
Merge remote-tracking branch 'upstream/master' into prototyping-compare

Commit:
Merge branch 'prototyping-compare' into prototyping

Commit:
review
